fullnode/fullnode_lazy.go

```diff
@@ -0,0 +1,235 @@
+	return &app, err
+}
+
+// GetSession:
+// - Uses the Shannon SDK to fetch a session for the (serviceID, appAddr) combination.
+// - Required to fulfill the FullNode interface.
+func (lfn *LazyFullNode) GetSession(
+	ctx context.Context,
+	serviceID sdk.ServiceID,
+	appAddr string,
+) (sessiontypes.Session, error) {
+	session, err := lfn.sessionClient.GetSession(
+		ctx,
+		appAddr,
+		string(serviceID),
+		0,
+	)
+
+	if err != nil {
+		return sessiontypes.Session{},
+			fmt.Errorf("GetSession: error getting the session for service %s app %s: %w",
+				serviceID, appAddr, err,
+			)
+	}
+
+	if session == nil {
+		return sessiontypes.Session{},
+			fmt.Errorf("GetSession: got nil session for service %s app %s: %w",
+				serviceID, appAddr, err,
+			)
+	}
+
+	return *session, nil
+}
+
+// ValidateRelayResponse validates the RelayResponse and verifies the supplier's signature.
+//
+// - Returns the RelayResponse, even if basic validation fails (may contain error reason).
+// - Verifies supplier's signature with the provided publicKeyFetcher.
+func (lfn *LazyFullNode) ValidateRelayResponse(
+	ctx context.Context,
+	supplierAddress sdk.SupplierAddress,
+	relayResponseBz []byte,
+) (*servicetypes.RelayResponse, error) {
+	relayResponse := &servicetypes.RelayResponse{}
+	if err := relayResponse.Unmarshal(relayResponseBz); err != nil {
+		return nil, err
+	}
+
+	if err := relayResponse.ValidateBasic(); err != nil {
+		// Even if the relay response is invalid, return it (may contain failure reason)
+		return relayResponse, err
+	}
+
+	supplierPubKey, err := lfn.GetAccountPubKey(
+		ctx,
+		string(supplierAddress),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// This can happen if a supplier has never been used (e.g. funded) onchain
+	if supplierPubKey == nil {
+		return nil, fmt.Errorf("ValidateRelayResponse: supplier public key is nil for address %s", string(supplierAddress))
+	}
+
+	if signatureErr := relayResponse.VerifySupplierOperatorSignature(supplierPubKey); signatureErr != nil {
+		return nil, signatureErr
+	}
+
+	return relayResponse, nil
+}
+
+// IsHealthy:
+// - Always returns true for a LazyFullNode.
+// - Required to fulfill the FullNode interface.
+func (lfn *LazyFullNode) IsHealthy() bool {
+	return true
+}
+
+// GetAccountPubKey returns the public key of the account with the given address.
+//
+// - Queries the account module using the gRPC query client.
+func (lfn *LazyFullNode) GetAccountPubKey(
+	ctx context.Context,
+	address string,
+) (pubKey cryptotypes.PubKey, err error) {
+
+	req := &accounttypes.QueryAccountRequest{Address: address}
+	res, err := lfn.accountClient.Account(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var fetchedAccount types.AccountI
+	if err = queryCodec.UnpackAny(res.Account, &fetchedAccount); err != nil {
+		return nil, err
+	}
+
+	return fetchedAccount.GetPubKey(), nil
+}
+
+func newSessionClient(config GRPCConfig) (*sdk.SessionClient, error) {
+	conn, err := connectGRPC(config)
```

Comment:
I feel like we should just set up one gRPC connect and reuse it everywhere.

No need to open more sockets on the fullnode -> it'll make it crash faster.

This way we can remove all of these helpers and just call `connectGRPC(config)` once.